File: python/tests/test_material_friction.py
import numpy as np
import scipy as sp

import find_ipctk
from ipctk import CollisionMesh, NormalCollisions, TangentialCollisions, FrictionPotential, BarrierPotential

def test_collision_mesh_material_ids():
    # Create a simple mesh
    vertices = np.array([
        [0, 0, 0],
        [1, 0, 0], 
        [0, 1, 0],
        [1, 1, 0]
    ], dtype=float)
    
    edges = np.array([
        [0, 1],
        [1, 3],
        [3, 2],
        [2, 0]
    ], dtype=int)
    
    # Create a collision mesh
    mesh = CollisionMesh(vertices, edges)
    
    # Set material IDs for vertices
    vertex_materials = np.array([0, 0, 1, 1], dtype=int)
    mesh.vertex_materials = vertex_materials
    
    # Check material IDs were set correctly
    assert mesh.has_material_ids()
    assert np.all(mesh.vertex_materials == vertex_materials)
    
    # Set a single material ID for everything
    mesh.set_single_material_id(2)
    assert np.all(mesh.vertex_materials == 2)
    assert np.all(mesh.edge_materials == 2)
    
    # Test NO_MATERIAL_ID case
    # Create a mesh without setting materials
    mesh2 = CollisionMesh(vertices, edges)
    assert not mesh2.has_material_ids()

def test_material_friction():
    # Create a simple scene with two objects of different materials
    vertices = np.array([
        [0, 0, 0],
        [1, 0, 0], 
        [0, 1, 0],
        [1, 1, 0]
    ], dtype=float)
    
    edges = np.array([
        [0, 1],
        [1, 3],
        [3, 2],
        [2, 0]
    ], dtype=int)
    
    # Create a collision mesh
    mesh = CollisionMesh(vertices, edges)
    
    # Set different materials for vertices
    vertex_materials = np.array([0, 0, 1, 1], dtype=int)
    mesh.vertex_materials = vertex_materials
    
    # Create a material-specific friction table
    friction_table = np.array([
        [0.3, 0.5],  # material 0 against 0 or 1
        [0.5, 0.7]   # material 1 against 0 or 1
    ])
    
    # Create normal collisions
    normal_collisions = NormalCollisions()
    normal_collisions.build(mesh, vertices, 0.1)
    
    # Create tangential collisions with material-specific friction
    tangential_collisions = TangentialCollisions()
    tangential_collisions.build(
        mesh, vertices, normal_collisions, 
        BarrierPotential(0.1), 100.0, 0.3  # default mu is 0.3
    )
    
    # Create a friction potential and set the material friction table
    friction_potential = FrictionPotential(1e-5)
    friction_potential.set_material_friction_table(friction_table)
    
    # Apply velocities to test friction
    velocities = np.zeros_like(vertices)
    velocities[2:4, 0] = 0.1  # Move vertices with material 1
    
    # Compute friction forces
    forces = friction_potential.force(
        tangential_collisions, mesh, vertices, 
        np.zeros_like(vertices), velocities,
        BarrierPotential(0.1), 100.0
    )
    
    # The forces should be non-zero where there are collisions
    assert np.linalg.norm(forces) > 0

File: tests/src/tests/friction/test_material_friction.cpp
#include <tests/friction/friction_data_generator.hpp>
#include <tests/utils.hpp>

#include <catch2/catch_test_macros.hpp>
#include <catch2/catch_approx.hpp>

#include <ipc/ipc.hpp>
#include <ipc/collisions/tangential/tangential_collisions.hpp>
#include <ipc/potentials/friction_potential.hpp>
#include <ipc/potentials/barrier_potential.hpp>

using namespace ipc;

TEST_CASE("Material-specific friction", "[friction][material]")
{
    // Create a simple scene with two objects of different materials
    Eigen::MatrixXd V(4, 3);
    V << 0, 0, 0, 
         1, 0, 0, 
         0, 1, 0,
         1, 1, 0; // 2D square in XY plane
         
    Eigen::MatrixXi E(4, 2);
    E << 0, 1,
         1, 3,
         3, 2,
         2, 0; // square edges
         
    CollisionMesh mesh(V, E);
    
    // Set material IDs - half of the vertices are material 0, half material 1
    Eigen::VectorXi vertex_materials(4);
    vertex_materials << 0, 0, 1, 1;
    mesh.set_vertex_materials(vertex_materials);
    CHECK(mesh.has_material_ids());
    
    // Edge materials derived from vertices
    Eigen::VectorXi edge_materials(4);
    edge_materials << 0, 0, 1, 1;
    mesh.set_edge_materials(edge_materials);
    
    // Simulate a vertex-vertex collision between material 0 and 1
    VertexVertexNormalCollision normal_collision(0, 2);
    normal_collision.material_id1 = mesh.vertex_material(0); // should be 0
    normal_collision.material_id2 = mesh.vertex_material(2); // should be 1
    CHECK(normal_collision.material_id1 == 0);
    CHECK(normal_collision.material_id2 == 1);
    
    // Create a friction collision from the normal collision
    VertexVertexTangentialCollision friction_collision(normal_collision);
    CHECK(friction_collision.material_id1 == 0);
    CHECK(friction_collision.material_id2 == 1);
    CHECK(friction_collision.has_material_ids());
    
    // Create a material-specific friction table
    Eigen::MatrixXd friction_table(2, 2);
    friction_table << 0.3, 0.5,
                      0.5, 0.7; // Different mu values for different material pairs
    
    // Set default friction coefficient
    friction_collision.mu = 0.4;
    
    // Create friction potential
    FrictionPotential friction_potential(0.001);
    
    // Test friction coefficient lookup
    double mu = friction_potential.get_friction_coefficient(friction_collision, friction_table);
    CHECK(mu == Catch::Approx(0.5)); // Should be the value for materials 0,1
    
    // Test with a different material combination
    friction_collision.material_id1 = 1;
    friction_collision.material_id2 = 1;
    mu = friction_potential.get_friction_coefficient(friction_collision, friction_table);
    CHECK(mu == Catch::Approx(0.7)); // Should be the value for materials 1,1
    
    // Test with invalid material IDs
    friction_collision.material_id1 = -1; // NO_MATERIAL_ID
    mu = friction_potential.get_friction_coefficient(friction_collision, friction_table);
    CHECK(mu == Catch::Approx(friction_collision.mu)); // Should fall back to collision's mu
    
    // Set valid material IDs but try with empty friction table
    friction_collision.material_id1 = 0;
    friction_collision.material_id2 = 1;
    mu = friction_potential.get_friction_coefficient(friction_collision, Eigen::MatrixXd(0, 0));
    CHECK(mu == Catch::Approx(friction_collision.mu)); // Should fall back to collision's mu
    
    // Test with material IDs out of range
    friction_collision.material_id1 = 5; // Out of range
    mu = friction_potential.get_friction_coefficient(friction_collision, friction_table);
    CHECK(mu == Catch::Approx(friction_collision.mu)); // Should fall back to collision's mu
}
